File: SignalServiceKit/tests/Storage/OWSIdentityManagerTests.m
//
//  Copyright (c) 2022 Open Whisper Systems. All rights reserved.
//

#import "MockSSKEnvironment.h"
#import "OWSIdentityManager.h"
#import "OWSRecipientIdentity.h"
#import "SSKBaseTestObjC.h"
#import "SSKEnvironment.h"
#import "TSAccountManager.h"
#import <Curve25519Kit/Curve25519.h>
#import <SignalCoreKit/Randomness.h>
#import <SignalServiceKit/SignalServiceKit-Swift.h>

@interface OWSIdentityManagerTests : SSKBaseTestObjC

@end

@implementation OWSIdentityManagerTests

- (void)setUp
{
    [super setUp];
    [self.tsAccountManager registerForTestsWithLocalNumber:@"+13235551234" uuid:[NSUUID UUID]];
}

- (void)tearDown
{
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (OWSIdentityManager *)identityManager
{
    return [OWSIdentityManager shared];
}

- (void)testNewEmptyKey
{
    NSData *newKey = [Randomness generateRandomBytes:32];
    SignalServiceAddress *address = [[SignalServiceAddress alloc] initWithPhoneNumber:@"+12223334444"];

    [self writeWithBlock:^(SDSAnyWriteTransaction *transaction) {
        __unused NSString *accountId = [OWSAccountIdFinder ensureAccountIdForAddress:address transaction:transaction];

        XCTAssert([self.identityManager isTrustedIdentityKey:newKey
                                                     address:address
                                                   direction:TSMessageDirectionOutgoing
                                                 transaction:transaction]);
        XCTAssert([self.identityManager isTrustedIdentityKey:newKey
                                                     address:address
                                                   direction:TSMessageDirectionIncoming
                                                 transaction:transaction]);
    }];
}

- (void)testAlreadyRegisteredKey
{
    NSData *newKey = [Randomness generateRandomBytes:32];
    SignalServiceAddress *address = [[SignalServiceAddress alloc] initWithPhoneNumber:@"+12223334444"];

    [self writeWithBlock:^(SDSAnyWriteTransaction *transaction) {
        [self.identityManager saveRemoteIdentity:newKey address:address transaction:transaction];

        XCTAssert([self.identityManager isTrustedIdentityKey:newKey
                                                     address:address
                                                   direction:TSMessageDirectionOutgoing
                                                 transaction:transaction]);
        XCTAssert([self.identityManager isTrustedIdentityKey:newKey
                                                     address:address
                                                   direction:TSMessageDirectionIncoming
                                                 transaction:transaction]);
    }];
}


- (void)testChangedKey
{
    NSData *originalKey = [Randomness generateRandomBytes:32];
    SignalServiceAddress *address = [[SignalServiceAddress alloc] initWithPhoneNumber:@"+12223334444"];

    [self writeWithBlock:^(SDSAnyWriteTransaction *transaction) {
        [self.identityManager saveRemoteIdentity:originalKey address:address transaction:transaction];

        XCTAssert([self.identityManager isTrustedIdentityKey:originalKey
                                                     address:address
                                                   direction:TSMessageDirectionOutgoing
                                                 transaction:transaction]);
        XCTAssert([self.identityManager isTrustedIdentityKey:originalKey
                                                     address:address
                                                   direction:TSMessageDirectionIncoming
                                                 transaction:transaction]);

        NSData *otherKey = [Randomness generateRandomBytes:32];

        XCTAssertFalse([self.identityManager isTrustedIdentityKey:otherKey
                                                          address:address
                                                        direction:TSMessageDirectionOutgoing
                                                      transaction:transaction]);
        XCTAssert([self.identityManager isTrustedIdentityKey:otherKey
                                                     address:address
                                                   direction:TSMessageDirectionIncoming
                                                 transaction:transaction]);
    }];
}

- (void)testIdentityKey
{
    ECKeyPair *newKey = [self.identityManager generateNewIdentityKeyForIdentity:OWSIdentityACI];
    XCTAssertEqual(newKey.publicKey.length, 32);

    ECKeyPair *pniKey = [self.identityManager generateNewIdentityKeyForIdentity:OWSIdentityPNI];
    XCTAssertEqual(pniKey.publicKey.length, 32);
    XCTAssertNotEqualObjects(pniKey.privateKey, newKey.privateKey);

    ECKeyPair *fetchedKey = [self.identityManager identityKeyPairForIdentity:OWSIdentityACI];
    XCTAssertEqualObjects(newKey.privateKey, fetchedKey.privateKey);

    ECKeyPair *fetchedPniKey = [self.identityManager identityKeyPairForIdentity:OWSIdentityPNI];
    XCTAssertEqualObjects(pniKey.privateKey, fetchedPniKey.privateKey);
}

@end

File: SignalServiceKit/tests/Storage/OWSIdentityManagerTests.swift
//
// Copyright 2022 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import LibSignalClient
import XCTest

@testable import SignalServiceKit

class OWSIdentityManagerTests: SSKBaseTest {
    private var identityManager: OWSIdentityManagerImpl { DependenciesBridge.shared.identityManager as! OWSIdentityManagerImpl }

    override func setUp() {
        super.setUp()
        SSKEnvironment.shared.databaseStorageRef.write { tx in
            (DependenciesBridge.shared.registrationStateChangeManager as! RegistrationStateChangeManagerImpl).registerForTests(
                localIdentifiers: .forUnitTests,
                tx: tx,
            )
        }
    }

    func testNewEmptyKey() throws {
        let newKey = IdentityKeyPair.generate().identityKey
        let aci = Aci.randomForTesting()
        try write { transaction in
            _ = DependenciesBridge.shared.recipientIdFinder.ensureRecipientUniqueId(for: aci, tx: transaction)
            XCTAssert(try identityManager.isTrustedIdentityKey(
                newKey,
                serviceId: aci,
                direction: .outgoing,
                tx: transaction,
            ))
            XCTAssert(try identityManager.isTrustedIdentityKey(
                newKey,
                serviceId: aci,
                direction: .incoming,
                tx: transaction,
            ))
        }
    }

    func testAlreadyRegisteredKey() throws {
        let newKey = IdentityKeyPair.generate().identityKey
        let aci = Aci.randomForTesting()
        try write { transaction in
            identityManager.saveIdentityKey(newKey, for: aci, tx: transaction)
            XCTAssert(try identityManager.isTrustedIdentityKey(
                newKey,
                serviceId: aci,
                direction: .outgoing,
                tx: transaction,
            ))
            XCTAssert(try identityManager.isTrustedIdentityKey(
                newKey,
                serviceId: aci,
                direction: .incoming,
                tx: transaction,
            ))
        }
    }

    func testChangedKey() throws {
        let originalKey = IdentityKeyPair.generate().identityKey
        let aci = Aci.randomForTesting()
        try write { transaction in
            identityManager.saveIdentityKey(originalKey, for: aci, tx: transaction)

            XCTAssert(try identityManager.isTrustedIdentityKey(
                originalKey,
                serviceId: aci,
                direction: .outgoing,
                tx: transaction,
            ))
            XCTAssert(try identityManager.isTrustedIdentityKey(
                originalKey,
                serviceId: aci,
                direction: .incoming,
                tx: transaction,
            ))

            let otherKey = IdentityKeyPair.generate().identityKey

            XCTAssertThrowsError(try identityManager.isTrustedIdentityKey(
                otherKey,
                serviceId: aci,
                direction: .outgoing,
                tx: transaction,
            ), "", { error in
                switch error {
                case IdentityManagerError.identityKeyMismatchForOutgoingMessage:
                    // This is fine.
                    break
                default:
                    XCTFail("Threw the wrong type of error.")
                }
            })
            XCTAssert(try identityManager.isTrustedIdentityKey(
                otherKey,
                serviceId: aci,
                direction: .incoming,
                tx: transaction,
            ))
        }
    }

    func testIdentityKey() {
        let identityManager = DependenciesBridge.shared.identityManager

        let newKey = identityManager.generateAndPersistNewIdentityKey(for: .aci)
        XCTAssertEqual(newKey.publicKey.count, 32)

        let pniKey = identityManager.generateAndPersistNewIdentityKey(for: .pni)
        XCTAssertEqual(pniKey.publicKey.count, 32)
        XCTAssertNotEqual(pniKey.privateKey, newKey.privateKey)

        let fetchedKey = SSKEnvironment.shared.databaseStorageRef.read { tx in identityManager.identityKeyPair(for: .aci, tx: tx)! }
        XCTAssertEqual(newKey.privateKey, fetchedKey.privateKey)

        let fetchedPniKey = SSKEnvironment.shared.databaseStorageRef.read { tx in identityManager.identityKeyPair(for: .pni, tx: tx)! }
        XCTAssertEqual(pniKey.privateKey, fetchedPniKey.privateKey)
    }

    func testShouldSharePhoneNumber() {
        let aliceAci = Aci.randomForTesting()
        let bobAci = Aci.randomForTesting()

        write { transaction in
            // {}
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: aliceAci, tx: transaction))
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: bobAci, tx: transaction))

            // {Alice}
            identityManager.setShouldSharePhoneNumber(with: aliceAci, tx: transaction)
            XCTAssertTrue(identityManager.shouldSharePhoneNumber(with: aliceAci, tx: transaction))
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: bobAci, tx: transaction))

            // {Alice}; redundant set shouldn't change anything.
            identityManager.setShouldSharePhoneNumber(with: aliceAci, tx: transaction)
            XCTAssertTrue(identityManager.shouldSharePhoneNumber(with: aliceAci, tx: transaction))
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: bobAci, tx: transaction))

            // {Alice, Bob}
            identityManager.setShouldSharePhoneNumber(with: bobAci, tx: transaction)
            XCTAssertTrue(identityManager.shouldSharePhoneNumber(with: aliceAci, tx: transaction))
            XCTAssertTrue(identityManager.shouldSharePhoneNumber(with: bobAci, tx: transaction))

            // {Bob}
            identityManager.clearShouldSharePhoneNumber(with: aliceAci, tx: transaction)
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: aliceAci, tx: transaction))
            XCTAssertTrue(identityManager.shouldSharePhoneNumber(with: bobAci, tx: transaction))

            // {Bob}; redundant clear shouldn't change anything.
            identityManager.clearShouldSharePhoneNumber(with: aliceAci, tx: transaction)
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: aliceAci, tx: transaction))
            XCTAssertTrue(identityManager.shouldSharePhoneNumber(with: bobAci, tx: transaction))

            // {Alice, Bob}
            identityManager.setShouldSharePhoneNumber(with: aliceAci, tx: transaction)
            // {}
            identityManager.clearShouldSharePhoneNumberForEveryone(tx: transaction)
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: aliceAci, tx: transaction))
            XCTAssertFalse(identityManager.shouldSharePhoneNumber(with: bobAci, tx: transaction))
        }
    }
}
